experimental/oak_baremetal_channel/src/client.rs

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{InvocationChannel, Message, MessageId, String, Vec};
+use ciborium_io::{Read, Write};
+
+pub struct ClientChannelHandle<T: Read + Write> {
+    inner: InvocationChannel<T>,
+}
+
+impl<T> ClientChannelHandle<T>
+where
+    T: Read<Error = anyhow::Error> + Write<Error = anyhow::Error>,
+{
+    pub fn new(socket: T) -> Self {
+        Self {
+            inner: InvocationChannel::new(socket),
+        }
+    }
+    pub fn write_request(&mut self, request: Message) -> anyhow::Result<()> {
+        self.inner.write_message(request)
+    }
+    pub fn read_response(&mut self) -> anyhow::Result<Message> {
+        let message = self.inner.read_message()?;
+        Ok(message)
+    }
+}
+
+#[derive(Default)]
+pub struct RequestEncoder {
+    message_id_counter: MessageIdCounter,
+}
+
+impl RequestEncoder {
+    pub fn encode_request(&mut self, request: oak_idl::Request) -> Message {
+        let message_id = self.message_id_counter.next_message_id();
+        Message {
+            message_id,
+            method_or_status: request.method_id,
+            body: request.body.to_vec(),
+        }
+    }
+}
+
+#[derive(Default)]
+struct MessageIdCounter {
+    next_message_id: MessageId,
+}
+
+impl MessageIdCounter {
+    fn next_message_id(&mut self) -> MessageId {
+        let next_message_id = self.next_message_id;
+        let _ = self.next_message_id.overflowing_add(1);
```

Comment:
I don't think this works. If I understand the documentation correctly `overflowing_add` returns a tuple which includes the sum, but does not increment the message id.

Also, if the code does not check whether it was overflowing or not it's probably better to use `wrapping_add`, or just use `Wrapping<u32>` as the type which would always wrap.